src/detector/yin.rs

```diff
@@ -0,0 +1,80 @@
+use crate::detector::internals::pitch_from_peaks;
+use crate::detector::internals::DetectorInternals;
+use crate::detector::internals::Pitch;
+use crate::detector::PitchDetector;
+use crate::float::Float;
+use crate::utils::buffer::square_sum;
+use crate::utils::peak::PeakCorrection;
+
+use super::internals::{windowed_square_error, yin_normalize_square_error};
+
+/// Pitch detection based on the YIN algorithm. See http://recherche.ircam.fr/equipes/pcm/cheveign/ps/2002_JASA_YIN_proof.pdf
+pub struct YINDetector<T>
+where
+    T: Float + std::iter::Sum,
+{
+    internals: DetectorInternals<T>,
+}
+
+impl<T> YINDetector<T>
+where
+    T: Float + std::iter::Sum,
+{
+    pub fn new(size: usize, padding: usize) -> Self {
+        let internals = DetectorInternals::new(1, 3, size, padding);
+        YINDetector { internals }
+    }
+}
+
+impl<T> PitchDetector<T> for YINDetector<T>
+where
+    T: Float + std::iter::Sum,
+{
+    fn get_pitch(
+        &mut self,
+        signal: &[T],
+        sample_rate: usize,
+        power_threshold: T,
+        clarity_threshold: T,
+    ) -> Option<Pitch<T>> {
+        // The YIN paper uses 0.1 as a threshold; TarsosDSP uses 0.2. `threshold` is not quite
+        // the same thing as 1 - clarity, but it should be close enough.
+        let threshold = T::one() - clarity_threshold;
+        let window_size = signal.len() / 2;
+
+        assert_eq!(signal.len(), self.internals.size);
+        assert!(
+            self.internals.has_sufficient_buffers(1, 3),
+            "YINDetector requires at least 1 real and 3 complex buffers"
+        );
+
+        if square_sum(signal) < power_threshold {
+            return None;
+        }
+
+        let mut iter = self.internals.complex_buffers.iter_mut();
+        let scratch1 = iter.next().unwrap();
+        let scratch2 = iter.next().unwrap();
+        let scratch3 = iter.next().unwrap();
+
+        let mut iter = self.internals.real_buffers.iter_mut();
+        let result = &mut iter.next().unwrap()[..window_size];
+
+        // STEP 2: Calculate the difference function, d_t.
+        windowed_square_error(signal, window_size, (scratch1, scratch2, scratch3), result);
+
+        // STEP 3: Calculate the cumulative mean normalized difference function, d_t'.
+        yin_normalize_square_error(result);
+
+        // STEP 4: The absolute threshold. We want the first dip below `threshold`.
+        // The YIN paper looks for minimum peaks. Since `pitch_from_peaks` looks
+        // for maximums, we take this opportunity to invert the signal.
+        result.iter_mut().for_each(|val| *val = threshold - *val);
```

Comment:
To test out the PR a little deeper, I updated the web app to be able to use it, and I noticed that `YIN` returns a negative clarity, which I suspect may be coming from this inversion.

Maybe we can adapt `pitch_from_peaks` to detect maxima/minima on demand, avoiding the need to invert the result.

Reply:
Did the numbers look sensible other than being negative? The YIN algorithm doesn't have the equivalent of a clarity...

The YIN algorithm looks for minimums between `threshold` crossings, so we could have a threshold and direction that we pass to `pitch_from_peaks`, but I don't see a way to detect such a thing automatically...Did you have ideas?

Perhaps there should be a simplified API for `detect_pitch` that just returns `Option<Pitch>` without a clarity, and then each algorithm has a `detect_pitch_with_details` that returns a type with whatever extra details it may have, like clarity or confidence, or whatnot.

Reply:
I thought about it a little, and I think we can still save some useful information for the clarity of the peak.

We can skip the `result = threshold - result` step, and instead call `pitch_from_peaks` with parameters `pitch_from_peaks(-result, sample_rate, -threshold)`.

Finally, before returning the `Pitch` struct, its clarity should be inverted.

I'm not sure if I thought it through, does it make sense to you?

Reply:
You're suggesting adding a fourth parameter to `pitch_from_peaks`?